src/daft-core/src/array/ops/binary.rs

```diff
@@ -249,6 +250,19 @@ impl BinaryArray {
 
         Ok(Self::from((self.name(), Box::new(arrow_result))))
     }
+
+    pub fn decode<Decoder>(&self, decoder: Decoder) -> DaftResult<Utf8Array>
+    where
+        Decoder: Fn(&[u8]) -> DaftResult<Cow<'_, str>>,
+    {
+        let arr = self.as_arrow();
+        let res = arr
+            .iter()
+            .map(|val| val.map_or(Ok(None), |s| decoder(s).map(Some)))
+            .collect::<DaftResult<arrow2::array::Utf8Array<i64>>>()?
+            .with_validity(arr.validity().cloned());
```

Comment:
perf: 

- This kind of operation would greatly benefit from parallelization. 
- lets pass in a `scratch` buffer in to `decoder` so we're not reallocating a buffer in this hot loop. 

```rs
let mut scratch = Vec::new();
// ...
decoder(s, &mut scratch)
```

- the `.iter` and `.collect` methods aren't highly optimized. Considering this could easily process large chunks of data, it'd be better to manually decode the `arr`'s values and reconstruct the array manually instead of using the iterator methods.